Error building on Linux

_From @sebastienros on January 8, 2018 17:33_

On Linux only, when building a specific project on Linux with the latest SDK, I get an exception that originates from msbuild.

When using an older SDK the error goes away.

Current installed SDK: `2.2.0-preview1-007866`
SDK that doesn't repro the error: `2.2.0-preview1-007522`

Project to reproduce the issue: https://github.com/aspnet/benchmarks/tree/dev/src/Benchmarks

Logged error
```
Errors
    /tmp/tmpYHoYVk.tmp/l5visbph.cxt/sdk/2.2.0-preview1-007866/Roslyn/Microsoft.CSharp.Core.targets(84,5): error MSB3883: Unexpected exception:  [/tmp/tmpYHoYVk.tmp/nshuithb.5nm/benchmarks/src/Benchmarks/Benchmarks.csproj]
    /tmp/tmpYHoYVk.tmp/l5visbph.cxt/sdk/2.2.0-preview1-007866/Roslyn/Microsoft.CSharp.Core.targets(84,5): System.AggregateException: One or more errors occurred. (The system cannot open the device or file specified) ---> System.IO.IOException: The system cannot open the device or file specified
   at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew)
   at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.<RunServerCompilationCore>d__7.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands)
---> (Inner Exception #0) System.IO.IOException: The system cannot open the device or file specified
   at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew)
   at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.<RunServerCompilationCore>d__7.MoveNext()<--- [/tmp/tmpYHoYVk.tmp/nshuithb.5nm/benchmarks/src/Benchmarks/Benchmarks.csproj]

```

I also join the full `msbuild.log`






_Copied from original issue: dotnet/cli#8344_

FWIW, I just hit a build error with a very similar exception. This was with SDK 2.1.300-preview2-008012

```
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error MSB3883: Unexpected exception:  [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error : System.AggregateException: One or more errors occurred. (The system cannot open the device or file specified) ---> System.IO.IOException: The system cannot open the device or file specified [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew) [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.<RunServerCompilationCore>d__7.MoveNext() [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    --- End of inner exception stack trace --- [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken) [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken) [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error : ---> (Inner Exception #0) System.IO.IOException: The system cannot open the device or file specified [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew) [~project~]
/root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.<RunServerCompilationCore>d__7.MoveNext()<--- [~project~]
```

I will try again on my side with the latest build, probably on Tuesday.
Unless someone looking at this issue is willing to try to repro the issue
with the project I linked to.

On Feb 17, 2018 1:48 PM, "Frederik Carlier" <notifications@github.com>
wrote:

> FWIW, I just hit a build error with a very similar exception. This was
> with SDK 2.1.300-preview2-008012
>
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error MSB3883: Unexpected exception:  [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error : System.AggregateException: One or more errors occurred. (The system cannot open the device or file specified) ---> System.IO.IOException: The system cannot open the device or file specified [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew) [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.<RunServerCompilationCore>d__7.MoveNext() [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    --- End of inner exception stack trace --- [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken) [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken) [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error : ---> (Inner Exception #0) System.IO.IOException: The system cannot open the device or file specified [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew) [~project~]
> /root/dotnet-latest/sdk/2.1.300-preview2-008012/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.<RunServerCompilationCore>d__7.MoveNext()<--- [~project~]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/roslyn/issues/24124#issuecomment-366473922>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABHJ7Xa8E5VV0aWX1IFOcwdWpZVLCdalks5tV0kzgaJpZM4RYIqU>
> .
>

Tried again today, here is the full log

```
[07:09:40.339] Downloading https://raw.githubusercontent.com/dotnet/cli/master/scripts/obtain/dotnet-install.sh
[07:09:40.559] Downloading https://raw.githubusercontent.com/dotnet/cli/master/scripts/obtain/dotnet-install.ps1
[07:09:40.568] Installing dotnet runtimes and sdk
[07:09:40.568] Downloading https://raw.githubusercontent.com/aspnet/BuildTools/dev/files/KoreBuild/config/sdk.version
[07:09:40.571] Detecting latest SDK version: 2.1.300-preview2-008251
[07:09:40.572] Downloading https://raw.githubusercontent.com/aspnet/Universe/dev/build/dependencies.props
[07:09:40.579] Detecting Universe Coherence runtime version: 2.1.0-preview2-26225-03
[07:09:40.579] Writing global.json with content: { "sdk": { "version": "2.1.300-preview2-008251" } }
[07:09:40.583] Downloading https://dotnet.myget.org/F/aspnetcore-dev/api/v3/registration1/aspnetcoreruntime/index.json
[07:09:40.900] Detecting latest runtime version: 2.1.0-preview2-30269
[07:09:40.900] /usr/bin/env bash dotnet-install.sh --channel Current
[07:09:41.455] [/usr/bin/env bash dotnet-install.sh --channel Current] dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.3/dotnet-sdk-2.1.3-linux-x64.tar.gz
[07:09:42.817] [/usr/bin/env bash dotnet-install.sh --channel Current] dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.3/dotnet-sdk-2.1.3-linux-x64.tar.gz
[07:09:47.366] [/usr/bin/env bash dotnet-install.sh --channel Current] dotnet-install: Adding to current process PATH: `/tmp/tmpWFqur1.tmp/njmiktqa.twj`. Note: This change will be visible only when sourcing script.
[07:09:47.366] [/usr/bin/env bash dotnet-install.sh --channel Current] dotnet-install: Installation finished successfully.
[07:09:47.366] [/usr/bin/env bash dotnet-install.sh --channel Current]
[07:09:47.366] [/usr/bin/env bash dotnet-install.sh --channel Current]
[07:09:47.452] /usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251
[07:09:47.566] [/usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251] dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.300-preview2-008251/dotnet-sdk-2.1.300-preview2-008251-linux-x64.tar.gz
[07:09:48.921] [/usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251] dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.300-preview2-008251/dotnet-sdk-2.1.300-preview2-008251-linux-x64.tar.gz
[07:09:51.263] [/usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251] dotnet-install: Adding to current process PATH: `/tmp/tmpWFqur1.tmp/njmiktqa.twj`. Note: This change will be visible only when sourcing script.
[07:09:51.263] [/usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251] dotnet-install: Installation finished successfully.
[07:09:51.263] [/usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251]
[07:09:51.263] [/usr/bin/env bash dotnet-install.sh --version 2.1.300-preview2-008251]
[07:09:51.291] /usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path
[07:09:51.400] [/usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path] dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/2.1.0-preview2-26225-03/dotnet-runtime-2.1.0-preview2-26225-03-linux-x64.tar.gz
[07:09:51.766] [/usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path] dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/2.1.0-preview2-26225-03/dotnet-runtime-2.1.0-preview2-26225-03-linux-x64.tar.gz
[07:09:52.365] [/usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path] dotnet-install: Binaries of dotnet can be found in /tmp/tmpWFqur1.tmp/njmiktqa.twj
[07:09:52.366] [/usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path] dotnet-install: Installation finished successfully.
[07:09:52.366] [/usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path]
[07:09:52.366] [/usr/bin/env bash dotnet-install.sh --version 2.1.0-preview2-26225-03 --runtime dotnet --no-path]
[07:09:52.428] /tmp/tmpWFqur1.tmp/njmiktqa.twj/dotnet restore /p:VersionSuffix=zzzzz-99999 /p:BenchmarksAspNetCoreVersion=2.1-* /p:BenchmarksNETStandardImplicitPackageVersion=2.1-* /p:BenchmarksNETCoreAppImplicitPackageVersion=2.1-* /p:BenchmarksRuntimeFrameworkVersion=2.1.0-preview2-26225-03 /p:BenchmarksTargetFramework=netcoreapp2.1
[07:09:53.367] Restoring packages for /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj...
[07:09:56.123] Installing System.Xml.XPath 4.3.0.
[07:09:56.136] Installing System.Buffers 4.0.0.
[07:09:56.137] Installing runtime.native.System.IO.Compression 4.1.0.
[07:09:56.137] Installing System.Threading.Overlapped 4.3.0.
[07:09:56.137] Installing System.Security.Cryptography.Csp 4.0.0.
[07:09:56.138] Installing System.Security.Cryptography.Cng 4.2.0.
[07:09:56.138] Installing runtime.native.System.Net.Http 4.0.1.
[07:09:56.139] Installing System.Threading.Tasks.Parallel 4.3.0.
[07:09:56.139] Installing System.Text.Encoding.CodePages 4.3.0.
[07:09:56.139] Installing System.Reflection.Metadata 1.4.2.
[07:09:56.346] Installing System.Xml.XPath.XDocument 4.3.0.
[07:09:56.347] Installing System.Diagnostics.FileVersionInfo 4.3.0.
[07:09:56.347] Installing System.Collections.Immutable 1.3.1.
[07:09:56.347] Installing System.ValueTuple 4.3.0.
[07:09:56.347] Installing System.Diagnostics.StackTrace 4.3.0.
[07:09:56.347] Installing Microsoft.CodeAnalysis.Analyzers 1.1.0.
[07:09:56.347] Installing Microsoft.AspNetCore.Diagnostics 2.1.0-preview2-30268.
[07:09:56.347] Installing Microsoft.AspNetCore.StaticFiles 2.1.0-preview2-30268.
[07:09:56.347] Installing Microsoft.AspNetCore.ResponseCaching 2.1.0-preview2-30268.
[07:09:56.347] Installing Microsoft.Extensions.Configuration.CommandLine 2.1.0-preview2-30268.
[07:09:56.347] Installing System.Globalization 4.0.11-rc2-24027.
[07:09:56.347] Installing Microsoft.AspNetCore.Server.HttpSys 2.1.0-preview2-30268.
[07:09:56.347] Installing Microsoft.NETCore.App 2.1.0-preview2-26225-03.
[07:09:56.347] Installing Microsoft.Extensions.Logging.Console 2.1.0-preview2-prkrishn-test-16441.
[07:09:56.347] Installing System.Data.SqlClient 4.4.2.
[07:09:56.347] Installing Npgsql.EntityFrameworkCore.PostgreSQL 2.0.0.
[07:09:56.347] Installing Microsoft.EntityFrameworkCore.SqlServer 2.1.0-preview2-30268.
[07:09:56.347] Installing Dapper 1.50.4.
[07:09:56.347] Installing Npgsql 3.2.7.
[07:09:56.347] Installing Pomelo.EntityFrameworkCore.MySql 2.0.0.
[07:09:56.347] Installing Jil 2.15.4.
[07:09:56.347] Installing MySqlConnector 0.36.1.
[07:09:56.347] Installing Microsoft.Extensions.Configuration.Json 2.1.0-preview2-30268.
[07:09:56.347] Installing Microsoft.AspNetCore.Mvc.Razor.ViewCompilation 2.1.0-preview2-30268.
[07:09:56.347] Installing Microsoft.AspNetCore.SignalR.MsgPack 1.0.0-preview2-30268.
[07:09:56.347] Installing Microsoft.AspNetCore.SignalR 1.0.0-preview2-30268.
[07:09:56.347] Installing MongoDB.Driver 2.5.0.
[07:09:56.347] Installing Microsoft.Extensions.Logging.Configuration 2.1.0-preview2-prkrishn-test-16441.
[07:09:56.347] Installing Microsoft.Extensions.Configuration.Abstractions 2.1.0-preview2-30131.
[07:09:56.347] Installing Microsoft.Extensions.Logging 2.1.0-preview2-prkrishn-test-16441.
[07:09:56.347] Installing Microsoft.NETCore.Targets 2.0.0.
[07:09:56.347] Installing Microsoft.NETCore.Platforms 2.1.0-preview2-26224-02.
[07:09:56.347] Installing Microsoft.NETCore.DotNetHostPolicy 2.1.0-preview2-26225-03.
[07:09:56.347] Installing NETStandard.Library 2.0.1.
[07:09:56.347] Installing System.Runtime.InteropServices.PInvoke 4.0.0-rc2-24027.
[07:09:56.347] Installing System.Security.Principal.Windows 4.4.0.
[07:09:56.347] Installing runtime.native.System.Data.SqlClient.sni 4.4.0.
[07:09:56.347] Installing Microsoft.Win32.Registry 4.4.0.
[07:09:56.347] Installing System.Text.Encoding.CodePages 4.4.0.
[07:09:56.347] Installing Microsoft.Extensions.DependencyInjection.Abstractions 2.0.0.
[07:09:56.347] Installing System.Runtime.Handles 4.0.1-rc2-24027.
[07:09:56.347] Installing Microsoft.Extensions.DependencyInjection 2.0.0.
[07:09:56.347] Installing Microsoft.EntityFrameworkCore.Relational 2.0.0.
[07:09:56.347] Installing Microsoft.EntityFrameworkCore 2.0.0.
[07:09:56.347] Installing Microsoft.EntityFrameworkCore.Relational 2.1.0-preview2-30268.
[07:09:56.347] Installing System.Reflection.TypeExtensions 4.4.0.
[07:09:56.347] Installing System.Reflection.Emit.Lightweight 4.3.0.
[07:09:56.347] Installing System.Threading.Tasks.Extensions 4.3.0.
[07:09:56.347] Installing System.Threading.Tasks 4.0.11-rc2-24027.
[07:09:56.347] Installing Microsoft.EntityFrameworkCore.Design 2.0.0.
[07:09:56.347] Installing Pomelo.JsonObject 1.1.1.
[07:09:56.347] Installing Sigil 4.7.0.
[07:09:56.347] Installing System.Runtime.Serialization.Primitives 4.3.0.
[07:09:56.347] Installing System.Collections.NonGeneric 4.3.0.
[07:09:56.348] Installing NETStandard.Library 1.6.1.
[07:09:56.348] Installing System.ComponentModel 4.3.0.
[07:09:56.348] Installing System.Reflection 4.3.0.
[07:09:56.348] Installing System.ComponentModel.TypeConverter 4.3.0.
[07:09:56.348] Installing Microsoft.CSharp 4.3.0.
[07:09:56.348] Installing System.Threading.Tasks.Extensions 4.0.0.
[07:09:56.348] Installing System.Dynamic.Runtime 4.3.0.
[07:09:56.348] Installing System.Collections 4.3.0.
[07:09:56.348] Installing System.Buffers 4.3.0.
[07:09:56.348] Installing System.Net.Http 4.1.0.
[07:09:56.350] Installing System.Net.NameResolution 4.3.0.
[07:09:56.351] Installing System.Net.Security 4.3.0.
[07:09:56.351] Installing System.Security.Cryptography.Algorithms 4.3.0.
[07:09:56.351] Installing System.Data.Common 4.3.0.
[07:09:56.351] Installing System.Diagnostics.Process 4.3.0.
[07:09:56.351] Installing Newtonsoft.Json 10.0.1.
[07:09:56.351] Installing Microsoft.AspNetCore.SignalR.Protocols.MsgPack 1.0.0-preview2-30268.
[07:09:56.351] Installing Microsoft.AspNetCore.SignalR.Core 1.0.0-preview2-30268.
[07:09:56.351] Installing Microsoft.AspNetCore.Sockets.Http 1.0.0-preview2-30268.
[07:09:56.351] Installing System.ComponentModel.TypeConverter 4.1.0.
[07:09:56.351] Installing System.Linq.Queryable 4.0.1.
[07:09:56.351] Installing MongoDB.Bson 2.5.0.
[07:09:56.351] Installing MongoDB.Driver.Core 2.5.0.
[07:09:56.351] Installing Microsoft.NETCore.Platforms 2.0.0.
[07:09:56.351] Installing Libuv 1.10.0.
[07:09:56.351] Installing Microsoft.Extensions.Primitives 2.1.0-preview2-30131.
[07:09:56.351] Installing Microsoft.Extensions.DependencyInjection.Abstractions 2.1.0-preview2-30131.
[07:09:56.351] Installing Microsoft.Extensions.Logging.Abstractions 2.1.0-preview2-prkrishn-test-16441.
[07:09:56.351] Installing System.IO.Compression.ZipFile 4.0.1.
[07:09:56.351] Installing Microsoft.Extensions.Configuration.Binder 2.1.0-preview2-30131.
[07:09:56.351] Installing Microsoft.Extensions.Options 2.1.0-preview2-30131.
[07:09:56.351] Installing Microsoft.NETCore.DotNetHostResolver 2.1.0-preview2-26225-03.
[07:09:56.351] Installing runtime.win-x64.runtime.native.System.Data.SqlClient.sni 4.4.0.
[07:09:56.351] Installing runtime.win-x86.runtime.native.System.Data.SqlClient.sni 4.4.0.
[07:09:56.351] Installing runtime.win-arm64.runtime.native.System.Data.SqlClient.sni 4.4.0.
[07:09:56.351] Installing System.Security.AccessControl 4.4.0.
[07:09:56.351] Installing Microsoft.CSharp 4.4.0.
[07:09:56.351] Installing Microsoft.Extensions.Configuration.Abstractions 2.0.0.
[07:09:56.351] Installing System.Diagnostics.DiagnosticSource 4.4.1.
[07:09:56.351] Installing System.ComponentModel.Annotations 4.4.0.
[07:09:56.351] Installing System.IO.Compression 4.1.0.
[07:09:56.351] Installing Microsoft.Extensions.Logging 2.0.0.
[07:09:56.351] Installing System.Collections.Immutable 1.4.0.
[07:09:56.351] Installing Microsoft.Extensions.Caching.Memory 2.0.0.
[07:09:56.352] Installing System.Interactive.Async 3.1.1.
[07:09:56.352] Installing Remotion.Linq 2.1.1.
[07:09:56.352] Installing Microsoft.EntityFrameworkCore 2.1.0-preview2-30268.
[07:09:56.352] Installing System.Reflection.Emit.ILGeneration 4.3.0.
[07:09:56.352] Installing System.Reflection.Primitives 4.3.0.
[07:09:56.352] Installing System.Runtime.InteropServices.RuntimeInformation 4.0.0.
[07:09:56.352] Installing System.Runtime 4.3.0.
[07:09:56.352] Installing Newtonsoft.Json 9.0.1.
[07:09:56.352] Installing System.Threading.Tasks 4.3.0.
[07:09:56.352] Installing System.Reflection.Emit.Lightweight 4.0.1-rc2-24027.
[07:09:56.352] Installing System.Reflection.Emit 4.0.1-rc2-24027.
[07:09:56.352] Installing System.Reflection.TypeExtensions 4.1.0-rc2-24027.
[07:09:56.352] Installing System.Reflection.Emit.ILGeneration 4.0.1-rc2-24027.
[07:09:56.352] Installing System.Reflection.Metadata 1.3.0-rc2-24027.
[07:09:56.352] Installing System.Resources.ResourceManager 4.3.0.
[07:09:56.353] Installing System.Threading.Timer 4.0.1.
[07:09:56.353] Installing System.Diagnostics.Debug 4.3.0.
[07:09:56.353] Installing System.Globalization 4.3.0.
[07:09:56.353] Installing System.Threading 4.3.0.
[07:09:56.353] Installing System.Runtime.Extensions 4.3.0.
[07:09:56.353] Installing Microsoft.NETCore.Platforms 1.1.0.
[07:09:56.353] Installing System.AppContext 4.3.0.
[07:09:56.353] Installing System.IO.FileSystem.Primitives 4.3.0.
[07:09:56.353] Installing System.IO.FileSystem 4.3.0.
[07:09:56.353] Installing System.Console 4.3.0.
[07:09:56.353] Installing System.Diagnostics.Tools 4.3.0.
[07:09:56.353] Installing Microsoft.Win32.Primitives 4.3.0.
[07:09:56.353] Installing System.Console 4.0.0.
[07:09:56.353] Installing System.Collections.Concurrent 4.3.0.
[07:09:56.370] Installing System.Text.Encoding.Extensions 4.3.0.
[07:09:56.371] Installing System.Runtime.Numerics 4.3.0.
[07:09:56.374] Installing System.IO.Compression.ZipFile 4.3.0.
[07:09:56.375] Installing System.Text.Encoding 4.3.0.
[07:09:56.375] Installing System.Threading.Timer 4.3.0.
[07:09:56.376] Installing System.Runtime.Handles 4.3.0.
[07:09:56.378] Installing System.Security.Cryptography.Primitives 4.3.0.
[07:09:56.379] Installing System.Linq 4.3.0.
[07:09:56.383] Installing System.IO 4.3.0.
[07:09:56.383] Installing System.Runtime.InteropServices.RuntimeInformation 4.3.0.
[07:09:56.385] Installing System.Globalization.Calendars 4.3.0.
[07:09:56.387] Installing System.Xml.XDocument 4.3.0.
[07:09:56.398] Installing System.Text.RegularExpressions 4.3.0.
[07:09:56.399] Installing System.Net.Sockets 4.3.0.
[07:09:56.399] Installing System.Net.Primitives 4.3.0.
[07:09:56.404] Installing System.IO.Compression 4.3.0.
[07:09:56.406] Installing System.Reflection.Extensions 4.3.0.
[07:09:56.408] Installing System.ObjectModel 4.3.0.
[07:09:56.419] Installing System.Security.Cryptography.Encoding 4.3.0.
[07:09:56.421] Installing System.Security.Cryptography.X509Certificates 4.3.0.
[07:09:56.427] Installing System.Xml.ReaderWriter 4.3.0.
[07:09:56.430] Installing System.Runtime.InteropServices 4.3.0.
[07:09:56.430] Installing System.Linq.Expressions 4.3.0.
[07:09:56.436] Installing System.Diagnostics.Tracing 4.3.0.
[07:09:56.439] Installing System.Net.Http 4.3.0.
[07:09:56.442] Installing System.Net.Sockets 4.1.0.
[07:09:56.442] Installing Microsoft.NETCore.Targets 1.1.0.
[07:09:56.445] Installing System.Reflection.TypeExtensions 4.3.0.
[07:09:56.449] Installing System.ComponentModel.Primitives 4.3.0.
[07:09:56.454] Installing System.Collections.Specialized 4.3.0.
[07:09:56.455] Installing System.Reflection.Emit 4.3.0.
[07:09:56.457] Installing runtime.native.System 4.3.0.
[07:09:56.459] Installing System.Security.Principal.Windows 4.3.0.
[07:09:56.461] Installing System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.461] Installing runtime.native.System.Net.Security 4.3.0.
[07:09:56.462] Installing System.Security.Principal 4.3.0.
[07:09:56.462] Installing System.Threading.ThreadPool 4.3.0.
[07:09:56.465] Installing runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.465] Installing System.Globalization.Extensions 4.3.0.
[07:09:56.476] Installing System.Security.Claims 4.3.0.
[07:09:56.481] Installing runtime.native.System.Security.Cryptography.Apple 4.3.0.
[07:09:56.483] Installing Microsoft.Win32.Registry 4.3.0.
[07:09:56.484] Installing System.Threading.Thread 4.3.0.
[07:09:56.487] Installing System.Xml.XmlDocument 4.3.0.
[07:09:56.490] Installing System.Runtime.Serialization.Formatters 4.3.0.
[07:09:56.490] Installing System.Globalization.Calendars 4.0.1.
[07:09:56.493] Installing Newtonsoft.Json.Bson 1.0.1.
[07:09:56.495] Installing MsgPack.Cli 1.0.0-rc.
[07:09:56.497] Installing Microsoft.AspNetCore.SignalR.Common 1.0.0-preview2-30268.
[07:09:56.498] Installing Microsoft.AspNetCore.Sockets.Abstractions 1.0.0-preview2-30268.
[07:09:56.498] Installing System.Threading.Channels 4.5.0-preview2-26224-02.
[07:09:56.502] Installing Microsoft.AspNetCore.Sockets.Common.Http 1.0.0-preview2-30268.
[07:09:56.504] Installing Microsoft.AspNetCore.Sockets 1.0.0-preview2-30268.
[07:09:56.505] Installing Microsoft.AspNetCore.WebSockets 2.1.0-preview2-30268.
[07:09:56.506] Installing System.Collections 4.0.11.
[07:09:56.507] Installing System.Linq 4.1.0.
[07:09:56.508] Installing System.Reflection.TypeExtensions 4.1.0.
[07:09:56.509] Installing System.Resources.ResourceManager 4.0.1.
[07:09:56.510] Installing System.Runtime.Extensions 4.1.0.
[07:09:56.511] Installing System.ComponentModel.Primitives 4.1.0.
[07:09:56.511] Installing System.ComponentModel 4.0.1.
[07:09:56.513] Installing System.Collections.NonGeneric 4.0.1.
[07:09:56.515] Installing System.AppContext 4.1.0.
[07:09:56.517] Installing System.Reflection.Extensions 4.0.1.
[07:09:56.518] Installing System.Globalization 4.0.11.
[07:09:56.526] Installing System.Collections.Specialized 4.0.1.
[07:09:56.529] Installing System.Threading 4.0.11.
[07:09:56.531] Installing System.Reflection 4.1.0.
[07:09:56.535] Installing System.Reflection.Primitives 4.0.1.
[07:09:56.545] Installing System.Runtime 4.1.0.
[07:09:56.547] Installing System.Diagnostics.Debug 4.0.11.
[07:09:56.549] Installing System.Linq.Expressions 4.1.0.
[07:09:56.551] Installing System.Reflection.Emit.Lightweight 4.0.1.
[07:09:56.551] Installing System.Dynamic.Runtime 4.0.11.
[07:09:56.551] Installing System.Diagnostics.Process 4.1.0.
[07:09:56.554] Installing System.Diagnostics.TraceSource 4.0.0.
[07:09:56.556] Installing System.Net.Security 4.0.0.
[07:09:56.563] Installing System.Security.SecureString 4.0.0.
[07:09:56.568] Installing DnsClient 1.0.7.
[07:09:56.568] Installing Microsoft.NETCore.Platforms 1.0.1.
[07:09:56.579] Installing Microsoft.Extensions.Configuration 2.1.0-preview2-30131.
[07:09:56.579] Installing System.Runtime.CompilerServices.Unsafe 4.5.0-preview2-26130-01.
[07:09:56.581] Installing Microsoft.NETCore.DotNetAppHost 2.1.0-preview2-26225-03.
[07:09:56.583] Installing Microsoft.Extensions.Primitives 2.0.0.
[07:09:56.584] Installing Microsoft.Extensions.Options 2.0.0.
[07:09:56.584] Installing Microsoft.Extensions.Caching.Abstractions 2.0.0.
[07:09:56.584] Installing Microsoft.Extensions.Logging.Abstractions 2.0.0.
[07:09:56.587] Installing NETStandard.Library 1.6.0.
[07:09:56.588] Installing System.ObjectModel 4.0.12.
[07:09:56.589] Installing Microsoft.EntityFrameworkCore.Attributes 2.1.0-preview2-30268.
[07:09:56.591] Installing Remotion.Linq 2.2.0.
[07:09:56.592] Installing System.IO 4.1.0.
[07:09:56.594] Installing System.Threading.Tasks 4.0.11.
[07:09:56.594] Installing System.Runtime.Serialization.Primitives 4.1.1.
[07:09:56.594] Installing System.Runtime.CompilerServices.Unsafe 4.4.0.
[07:09:56.598] Installing System.Text.RegularExpressions 4.1.0.
[07:09:56.599] Installing System.Text.Encoding.Extensions 4.0.11.
[07:09:56.600] Installing System.Xml.XDocument 4.0.11.
[07:09:56.606] Installing System.Xml.ReaderWriter 4.0.11.
[07:09:56.608] Installing System.Text.Encoding 4.0.11.
[07:09:56.611] Installing Microsoft.CSharp 4.0.1.
[07:09:56.613] Installing System.Reflection.Primitives 4.0.1-rc2-24027.
[07:09:56.616] Installing System.Runtime 4.1.0-rc2-24027.
[07:09:56.633] Installing System.IO 4.1.0-rc2-24027.
[07:09:56.715] Installing System.Diagnostics.Debug 4.0.11-rc2-24027.
[07:09:56.715] Installing System.Reflection.Extensions 4.0.1-rc2-24027.
[07:09:56.715] Installing System.Linq 4.1.0-rc2-24027.
[07:09:56.715] Installing System.Text.Encoding.Extensions 4.0.11-rc2-24027.
[07:09:56.715] Installing System.Runtime.InteropServices 4.1.0-rc2-24027.
[07:09:56.715] Installing System.Text.Encoding 4.0.11-rc2-24027.
[07:09:56.715] Installing System.Threading 4.0.11-rc2-24027.
[07:09:56.715] Installing System.Runtime.Extensions 4.1.0-rc2-24027.
[07:09:56.715] Installing System.Resources.ResourceManager 4.0.1-rc2-24027.
[07:09:56.715] Installing System.Collections.Immutable 1.2.0-rc2-24027.
[07:09:56.716] Installing runtime.native.System.IO.Compression 4.3.0.
[07:09:56.716] Installing runtime.native.System.Net.Http 4.3.0.
[07:09:56.716] Installing System.Security.Cryptography.Csp 4.3.0.
[07:09:56.716] Installing System.Security.Cryptography.Cng 4.3.0.
[07:09:56.716] Installing System.Diagnostics.DiagnosticSource 4.3.0.
[07:09:56.716] Installing runtime.osx.10.10-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.opensuse.42.1-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.fedora.24-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.fedora.23-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.rhel.7-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.ubuntu.16.10-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.opensuse.13.2-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.ubuntu.16.04-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.ubuntu.14.04-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.debian.8-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
[07:09:56.716] Installing runtime.osx.10.10-x64.runtime.native.System.Security.Cryptography.Apple 4.3.0.
[07:09:56.716] Installing System.Collections.Concurrent 4.0.12.
[07:09:56.716] Installing System.Runtime.InteropServices 4.1.0.
[07:09:56.716] Installing System.Diagnostics.Tools 4.0.1.
[07:09:56.716] Installing System.Diagnostics.Contracts 4.0.1.
[07:09:56.716] Installing System.Data.Common 4.1.0.
[07:09:56.716] Installing System.Numerics.Vectors 4.1.1.
[07:09:56.716] Installing System.Runtime.Numerics 4.0.1.
[07:09:56.716] Installing System.Net.WebSockets.WebSocketProtocol 4.5.0-preview2-26224-02.
[07:09:56.716] Installing Microsoft.NETCore.Targets 1.0.1.
[07:09:56.716] Installing System.Globalization.Extensions 4.0.1.
[07:09:56.716] Installing System.Reflection.Emit 4.0.1.
[07:09:56.716] Installing System.Reflection.Emit.ILGeneration 4.0.1.
[07:09:56.716] Installing Microsoft.Win32.Primitives 4.0.1.
[07:09:56.716] Installing runtime.native.System 4.0.0.
[07:09:56.716] Installing System.IO.FileSystem 4.0.1.
[07:09:56.716] Installing System.IO.FileSystem.Primitives 4.0.1.
[07:09:56.716] Installing System.Runtime.Handles 4.0.1.
[07:09:56.716] Installing Microsoft.Win32.Registry 4.0.0.
[07:09:56.716] Installing System.Threading.Thread 4.0.0.
[07:09:56.716] Installing System.Threading.ThreadPool 4.0.10.
[07:09:56.716] Installing runtime.native.System.Security.Cryptography 4.0.0.
[07:09:56.716] Installing System.Diagnostics.Tracing 4.1.0.
[07:09:56.716] Installing System.Net.Primitives 4.0.11.
[07:09:56.716] Installing System.Security.Cryptography.Encoding 4.0.0.
[07:09:56.716] Installing System.Security.Cryptography.Primitives 4.0.0.
[07:09:56.716] Installing runtime.native.System.Net.Security 4.0.1.
[07:09:56.716] Installing System.Security.Principal 4.0.1.
[07:09:56.716] Installing System.Security.Claims 4.0.1.
[07:09:56.716] Installing System.Security.Cryptography.Algorithms 4.2.0.
[07:09:56.716] Installing System.Security.Cryptography.OpenSsl 4.0.0.
[07:09:56.716] Installing System.Security.Cryptography.X509Certificates 4.1.0.
[07:09:56.716] Installing System.Net.NetworkInformation 4.3.0.
[07:09:57.012] Generating MSBuild file /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/obj/Benchmarks.csproj.nuget.g.props.
[07:09:57.014] Generating MSBuild file /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/obj/Benchmarks.csproj.nuget.g.targets.
[07:09:57.033] Restore completed in 3.67 sec for /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj.
[07:09:57.095] 
[07:09:57.095] 
[07:09:57.168] /tmp/tmpWFqur1.tmp/njmiktqa.twj/dotnet publish -c Release -o /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/published /p:BenchmarksAspNetCoreVersion=2.1-* /p:BenchmarksNETStandardImplicitPackageVersion=2.1-* /p:BenchmarksNETCoreAppImplicitPackageVersion=2.1-* /p:BenchmarksRuntimeFrameworkVersion=2.1.0-preview2-26225-03 /p:BenchmarksTargetFramework=netcoreapp2.1  /p:PublishWithAspNetCoreTargetManifest=false /p:MicrosoftNETPlatformLibrary=Microsoft.NETCore.App
[07:09:57.361] Microsoft (R) Build Engine version 15.6.82.30579 for .NET Core
[07:09:57.361] Copyright (C) Microsoft Corporation. All rights reserved.
[07:09:57.361] 
[07:09:58.204]   Restoring packages for /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj...
[07:09:59.055]   Restore completed in 848.36 ms for /tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj.
[07:09:59.592] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error MSB3883: Unexpected exception:  [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.596] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error : System.AggregateException: One or more errors occurred. (The system cannot open the device or file specified) ---> System.IO.IOException: The system cannot open the device or file specified [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.596] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew) [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.596] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.RunServerCompilationCore(RequestLanguage language, List`1 arguments, BuildPathsAlt buildPaths, String pipeName, String keepAlive, String libEnvVariable, Nullable`1 timeoutOverride, Func`3 tryCreateServerFunc, CancellationToken cancellationToken) [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    --- End of inner exception stack trace --- [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken) [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken) [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error : ---> (Inner Exception #0) System.IO.IOException: The system cannot open the device or file specified [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew) [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.597] /tmp/tmpWFqur1.tmp/njmiktqa.twj/sdk/2.1.300-preview2-008251/Roslyn/Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CommandLine.BuildServerConnection.RunServerCompilationCore(RequestLanguage language, List`1 arguments, BuildPathsAlt buildPaths, String pipeName, String keepAlive, String libEnvVariable, Nullable`1 timeoutOverride, Func`3 tryCreateServerFunc, CancellationToken cancellationToken)<--- [/tmp/tmpWFqur1.tmp/ispx1l5a.jms/benchmarks/src/Benchmarks/Benchmarks.csproj]
[07:09:59.650] 
```

--- a/src/Compilers/Server/VBCSCompilerTests/DesktopBuildClientTests.cs
+++ b/src/Compilers/Server/VBCSCompilerTests/DesktopBuildClientTests.cs
@@ -15,6 +15,7 @@ using Microsoft.CodeAnalysis.Test.Utilities;
 using Roslyn.Test.Utilities;
 using System.Threading;
 using System.IO.Pipes;
+using System.Security.AccessControl;
 
 namespace Microsoft.CodeAnalysis.CompilerServer.UnitTests
 {
@@ -144,6 +145,32 @@ namespace Microsoft.CodeAnalysis.CompilerServer.UnitTests
                     Assert.True(ranLocal);
                 }
             }
+
+#if NET461
+            [Fact]
+            public void TestMutexConstructorException()
+            {
+                using (var outer = new Mutex(initiallyOwned: true, name: BuildServerConnection.GetClientMutexName(_pipeName), out bool createdNew))
+                {
+                    Assert.True(createdNew);
+                    var mutexSecurity = outer.GetAccessControl();
+                    var user = Environment.UserDomainName + "\\" + Environment.UserName;
+                    mutexSecurity.AddAccessRule(new MutexAccessRule(user, MutexRights.FullControl, AccessControlType.Deny));
+                    outer.SetAccessControl(mutexSecurity);
+
+                    var ranLocal = false;
+                    var client = CreateClient(
+                        compileFunc: delegate
+                        {
+                            ranLocal = true;
+                            return 0;
+                        });
+                    var exitCode = client.RunCompilation(new[] { "/shared" }, _buildPaths).ExitCode;
+                    Assert.Equal(0, exitCode);
+                    Assert.True(ranLocal);
+                }
+            }
+#endif
 
             [Fact]
             public async Task ConnectToPipe()

--- a/src/Compilers/Shared/BuildServerConnection.cs
+++ b/src/Compilers/Shared/BuildServerConnection.cs
@@ -118,47 +118,62 @@ namespace Microsoft.CodeAnalysis.CommandLine
             var clientDir = buildPaths.ClientDirectory;
             var timeoutNewProcess = timeoutOverride ?? TimeOutMsNewProcess;
             var timeoutExistingProcess = timeoutOverride ?? TimeOutMsExistingProcess;
-            var clientMutexName = GetClientMutexName(pipeName);
             Task<NamedPipeClientStream> pipeTask = null;
-            using (var clientMutex = new Mutex(initiallyOwned: true,
-                                               name: clientMutexName,
-                                               createdNew: out var holdsMutex))
+            Mutex clientMutex = null;
+            var holdsMutex = false;
+            try
             {
                 try
                 {
-                    if (!holdsMutex)
-                    {
-                        try
-                        {
-                            holdsMutex = clientMutex.WaitOne(timeoutNewProcess);
+                    var clientMutexName = GetClientMutexName(pipeName);
+                    clientMutex = new Mutex(initiallyOwned: true, name: clientMutexName, out holdsMutex);
+                }
+                catch
+                {
+                    // The Mutex constructor can throw in certain cases. One specific example is docker containers
+                    // where the /tmp directory is restricted. In those cases there is no reliable way to execute
+                    // the server and we need to fall back to the command line.
+                    //
+                    // Example: https://github.com/dotnet/roslyn/issues/24124
+                    return new RejectedBuildResponse();
+                }
 
-                            if (!holdsMutex)
-                            {
-                                return new RejectedBuildResponse();
-                            }
-                        }
-                        catch (AbandonedMutexException)
+                if (!holdsMutex)
+                {
+                    try
+                    {
+                        holdsMutex = clientMutex.WaitOne(timeoutNewProcess);
+
+                        if (!holdsMutex)
                         {
-                            holdsMutex = true;
+                            return new RejectedBuildResponse();
                         }
                     }
-
-                    // Check for an already running server
-                    var serverMutexName = GetServerMutexName(pipeName);
-                    bool wasServerRunning = WasServerMutexOpen(serverMutexName);
-                    var timeout = wasServerRunning ? timeoutExistingProcess : timeoutNewProcess;
-
-                    if (wasServerRunning || tryCreateServerFunc(clientDir, pipeName))
+                    catch (AbandonedMutexException)
                     {
-                        pipeTask = TryConnectToServerAsync(pipeName, timeout, cancellationToken);
+                        holdsMutex = true;
                     }
                 }
-                finally
+
+                // Check for an already running server
+                var serverMutexName = GetServerMutexName(pipeName);
+                bool wasServerRunning = WasServerMutexOpen(serverMutexName);
+                var timeout = wasServerRunning ? timeoutExistingProcess : timeoutNewProcess;
+
+                if (wasServerRunning || tryCreateServerFunc(clientDir, pipeName))
+                {
+                    pipeTask = TryConnectToServerAsync(pipeName, timeout, cancellationToken);
+                }
+            }
+            finally
+            {
+                if (clientMutex != null)
                 {
                     if (holdsMutex)
                     {
                         clientMutex.ReleaseMutex();
                     }
+                    clientMutex.Dispose();
                 }
             }
 
@@ -555,12 +570,21 @@ namespace Microsoft.CodeAnalysis.CommandLine
 
         internal static bool WasServerMutexOpen(string mutexName)
         {
-            Mutex mutex;
-            var open = Mutex.TryOpenExisting(mutexName, out mutex);
-            if (open)
+            try
             {
-                mutex.Dispose();
-                return true;
+                Mutex mutex;
+                var open = Mutex.TryOpenExisting(mutexName, out mutex);
+                if (open)
+                {
+                    mutex.Dispose();
+                    return true;
+                }
+            }
+            catch
+            {
+                // In the case an exception occured trying to open the Mutex then 
+                // the assumption is that it's not open. 
+                return false;
             }
 
             return false;

